### ISSUE TITLE
A question for the wildcard certificate

### ISSUE THREAD

**Issue by colorful-sky** [libreswan/libreswan#363]
Hello Libreswan Team.

When I use Libreswan, I found that Libreswan supports wildcard certificates for id matching (ID_FQND type). For example, when a certificate (CN=abcde, SAN dns: *.org) issued by CA and its corresponding private key is added, the relevant device can connect to the IPsec server (with config rightid=@user1.org).

Although the wildcard certificate can facilitate server management, it may increase network security risks. A certificate can contain multiple SAN fields, and the validity period can be set for a long time (10 years) when applying.

In addition, users with the corresponding wildcard certificate can easily initiate identity spoofing attacks, or even man-in-the-middle attacks, while the deceived device cannot detect it at all.

According to RFC4945, "substring, wildcard, or regular expression matching MUST NOT be performed for this comparison." (ID_FQDN type ID matching, in 3.1.2 section)

I think it seems to be an implementation error and security risk, and should be taken seriously. 

Have you some ideas?

Thank you so much.

**Comment by colorful-sky** [comment]
Hi, paul.
Thanks for your quick response and answers.

> The client certificates are always issued by a CA that the VPN server trusts. In almost every case this is a private CA. For the VPN server certificate, this is often a certificate signed by a public CA, so it can be re-used as SSL-VPN certificate as well. These certificates are usually constrained by the CAB/Forum rules.
>
> For a private CA, yes you have all the power and it comes with great responsibility. Public CA's wont issue certs with a 13 month validity.

I agree with your these views. In my question, what I mean and what I am concerned about is the private CA, after all, this is the main cert issuer in the most  IPsec application scenario.

> So ensure your configuration and ID requirements are strict enough ?
>
> I'm not sure what the real risk is here. Perhaps you can explain this
> with mentioned which CA is trusted where and how you would think
> people deploy their connection configurations that are dangerous ?

These questions may be hard to answer.
What I described may be a bit exaggerated, because the security of certificate authentication mainly lies in the security of the private key, and the strict setting of ID can only play a certain complementary role.
Most users may not care whether the ID matching is reasonable and safe or not (for example, to ensure a smooth connection, set rightid=%any), but this does not seem to prevent the Libreswan project from working towards a more secure and compliant direction.

> It is not. That section means you can only have "*.something.somethingelse.
> You cannot have foo.*.bar or foo*.org or foo*.* And you cannot have regular
> expressions. We fully comply with these requirement,

Let us focus on here. Now, the crux of the question and what puzzles me is, which norm does this view come from? RFC4301, RFC6125, or other specifications?

There is such a description in RFC4301 section 4.4.3.1 : 

> "A DNS name may be fully qualified and thus match
>    exactly one name, e.g., foo.example.com.  Alternatively, the name may
>    encompass a group of peers by being partially specified, e.g., the
>    string ".example.com" could be used to match any DNS name ending in
>    these two domain name components." (4.4.3.1. PAD Entry IDs and Matching Rules)

But I think these content may correspond to the last paragraph in section 3.1.2 of RFC4945 and have little to do with the certificate matching : 

> "Implementations MAY support substring, wildcard, or regular
>    expression matching of the contents of ID to look up the policy in
>    the SPD, and such would be a matter of local security policy
>    configuration." (The last paragraph in 3.1.2 section, ID_FQDN)



There is such a description in RFC6125 section 6.4.3 : 

> "1.  The client SHOULD NOT attempt to match a presented identifier in
>        which the wildcard character comprises a label other than the
>        left-most label (e.g., do not match bar.*.example.net)." (6.4.3.  Checking of Wildcard Certificates)

But the content of RFC6125 is for TLS PKIX, not IPsec.

The problem of the wildcard certificate was discovered when I was studying and sorting out the IPsec protocol. I just want to figure out the answer to this question because it is related to my experiment report (^_^).

In addition, I just tried Strongswan 5.90 and got the opposite result from Libreswan, that is, the device with wildcard cert  (CN=abcde, SAN dns: *.org) can not connect with the Strongswan server. (with config rightid=@user1.org)

This is more confusing to me, which implementation is correct?
Can you answer my doubts? thank you very much!

**Comment by letoams** [comment]
On Fri, 11 Sep 2020, colorful-sky wrote:

> These questions may be hard to answer.
> What I described may be a bit exaggerated, because the security of certificate authentication mainly lies in the security of the private key, and the strict setting of ID can only play a certain
> complementary role.
> Most users may not care whether the ID matching is reasonable and safe or not (for example, to ensure a smooth connection, set rightid=%any), but this does not seem to prevent the Libreswan project from
> working towards a more secure and compliant direction.

Most private CA's don't issue anything with wildcards on it. The main
reason for seeing wildcards within the cert is for VPN server load
balancing. And deployments there are moving away from it because it
is causing issues for interoperability.

>       It is not. That section means you can only have ".something.somethingelse.
>       You cannot have foo..bar or foo*.org or foo*.* And you cannot have regular
>       expressions. We fully comply with these requirement,
> 
> Let us focus on here. Now, the crux of the question and what puzzles me is, which norm does this view come from? RFC4301, RFC6125, or other specifications?

Those are generic PKIX rules.

> There is such a description in RFC4301 section 4.4.3.1 :
>
>       "A DNS name may be fully qualified and thus match
>       exactly one name, e.g., foo.example.com. Alternatively, the name may
>       encompass a group of peers by being partially specified, e.g., the
>       string ".example.com" could be used to match any DNS name ending in
>       these two domain name components." (4.4.3.1. PAD Entry IDs and Matching Rules)

This talks about matching of a non-wildcarded certificate with
SubjectAltNames that do not contain wildcards. Eg for the server
to be able to match a certificate of SAN DNS:paul.developers.example.org
to its own grouping, eg to give all developers access to the development
lab by "matching" the connection for the lab server with
*.developers.org

> But I think these content may correspond to the last paragraph in section 3.1.2 of RFC4945 and have little to do with the certificate matching :
>
>       "Implementations MAY support substring, wildcard, or regular
>       expression matching of the contents of ID to look up the policy in
>       the SPD, and such would be a matter of local security policy
>       configuration." (The last paragraph in 3.1.2 section, ID_FQDN)

Right. Same here. the wildcard is for matching. The wildcard is not
being matched.

> There is such a description in RFC6125 section 6.4.3 :
>
>       "1. The client SHOULD NOT attempt to match a presented identifier in
>       which the wildcard character comprises a label other than the
>       left-most label (e.g., do not match bar.*.example.net)." (6.4.3. Checking of Wildcard Certificates)
> 
> But the content of RFC6125 is for TLS PKIX, not IPsec.

This just re-iterates PKIX rules. There can be only one '*' character.
It must be leftmost. It must constitute the entire label.

> The problem of the wildcard certificate was discovered when I was studying and sorting out the IPsec protocol. I just want to figure out the answer to this question because it is related to my experiment
> report (^_^).
> 
> In addition, I just tried Strongswan 5.90 and got the opposite result from Libreswan, that is, the device with wildcard cert (CN=abcde, SAN dns: *.org) can not connect with the Strongswan server. (with
> config rightid=@user1.org)
> 
> This is more confusing to me, which implementation is correct?

We have 200 test cases for various cert/x509 scenarios. You have to be a
bit more specific of you are asking. Please clarify the SAN entry and the
configured IKE ID's on initiator and responder.

In other words. Show the SAN and the leftid/rightid config of client and
server. Then we can talk about what should or shold not be accepted
and whether libreswan is doing something wrong.

Paul

**Comment by colorful-sky** [comment]
Hi, paul.

Thanks for your answers again! I show my test results as follows:

### Designed experiment

I wrote a python IKEv2 client program to ensure that I can accurately control the IDi Type and IDi Data sent each time. All test certificates are generated by Libreswan certutil. [(I put the test code, certificate and log in my repository )](https://github.com/colorful-sky/IKEv2-Client-for-ID-match-test)

Three machines participated in the test：

device A: Python IKEv2 Client, 192.168.0.1

device B: Libreswan Server, 192.168.0.136

device C: Strongswan Server, 192.168.0.128

#### test1: 

Python Client A ( with ID_FQDN: *.org and cert CN=abcde, SAN dns: *.org) to connect with Libreswan Server B (with config rightid=@user1.org)

Result: AUTHENTICATION succeeded

B: ipsec.conf

```
version 2.0
config setup
  virtual-private=%v4:10.0.0.0/8,%v4:192.168.0.0/16,%v4:172.16.0.0/12,%v4:!192.168.42.0/24,%v4:!192.168.43.0/24
  protostack=netkey
  interfaces=%defaultroute
  uniqueids=no
  plutodebug="all crypt"
  plutostderrlog=/usr/local/etc/libreswan.log

include /etc/ipsec.d/*.conf


conn ikev2-cert
  left=%defaultroute
  leftcert=192.168.0.136
  leftid=@192.168.0.136
  leftsendcert=always
  # leftsubnet=0.0.0.0/0
  leftrsasigkey=%cert
  right=192.168.0.1
  rightid="@user1.org"
  # rightaddresspool=192.168.43.10/24
  rightca=%same
  rightrsasigkey=%cert
  narrowing=yes
  dpddelay=30
  dpdtimeout=120
  dpdaction=clear
  auto=add
  ikev2=insist
  rekey=no
  pfs=no
  ike-frag=yes
  ike=3des-sha1;modp1024,
  phase2alg=aes_gcm-null,aes128-sha1,aes256-sha1,aes128-sha2,aes256-sha2,3des-sha1
  modecfgdns="8.8.8.8 8.8.4.4"
  encapsulation=yes
  mobike=no
```

B: libreswan_user1_org.log

```
Sep 13 20:56:51.637069: | checking for known CERT payloads
Sep 13 20:56:51.637072: | saving certificate of type 'X509_SIGNATURE'
Sep 13 20:56:51.638534: | decoded cert: CN=abcde
Sep 13 20:56:51.638547: |       #1 spent 1.48 milliseconds in find_and_verify_certs() calling decode_cert_payloads()
Sep 13 20:56:51.638551: | cert_issuer_has_current_crl: looking for a CRL issued by O=IKEv2 VPN,CN=IKEv2 VPN CA
Sep 13 20:56:51.638807: |       #1 spent 0.254 milliseconds in find_and_verify_certs() calling crl_update_check()
Sep 13 20:56:51.638813: | missing or expired CRL
Sep 13 20:56:51.638816: | crl_strict: 0, ocsp: 0, ocsp_strict: 0, ocsp_post: 0
Sep 13 20:56:51.638818: | verify_end_cert verifying CN=abcde using:
Sep 13 20:56:51.638820: |   trusted CA: O=IKEv2 VPN,CN=IKEv2 VPN CA
Sep 13 20:56:51.638822: | verify_end_cert trying profile TLS Client
Sep 13 20:56:51.639574: | certificate is valid (profile TLS Client)
Sep 13 20:56:51.639584: |       #1 spent 0.767 milliseconds in find_and_verify_certs() calling verify_end_cert()
Sep 13 20:56:51.639587: | delref root_certs@0x55b124e6a908(2->1) (in decode_certs() at x509.c:717)
Sep 13 20:56:51.639592: "ikev2-cert"[1] 192.168.0.1 #1: certificate verified OK: CN=abcde
Sep 13 20:56:51.639887: | get_pluto_gn_from_nss_cert: allocated pluto_gn 0x55b124e6ecf8
Sep 13 20:56:51.639895: | get_pluto_gn_from_nss_cert: allocated pluto_gn 0x55b124e7d0f8
Sep 13 20:56:51.639975: | unreference key: 0x55b124e73448 CN=abcde cnt 1--
Sep 13 20:56:51.640023: |       #1 spent 0.426 milliseconds in decode_certs() calling add_pubkey_from_nss_cert()
Sep 13 20:56:51.640028: |     #1 spent 6.22 milliseconds in decode_certs()
Sep 13 20:56:51.640034: | subjectAltname 'user1.org' matched '*.org' in certificate
Sep 13 20:56:51.640037: | X509: CERT and ID matches current connection
```

test2: 

```
Python Client A ( with ID_FQDN: *.org and cert CN=abcde, SAN dns: *.org) to connect with Strongswan Server C (with config rightid=@*.org)
```

Result: AUTHENTICATION succeeded

C: ipsec.conf

```
config setup

conn %default
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1

conn test
    keyexchange=ikev2
    ike=3des-sha1-modp1024!
    rekey=no
    left=192.168.0.128
    leftid=%any
    leftauth=pubkey
    leftsubnet=0.0.0.0/0
    leftcert=192.168.0.128.pem
    leftfirewall=yes
    rightsourceip=10.31.2.0/24
    right=192.168.0.1
    rightid="@*.org"
    rightauth=pubkey
    dpdaction=clear
    fragmentation=yes
    auto=add
```

C:strongswan_star_org.log

```
Sep 13 21:08:42 05[CFG]   using certificate "CN=abcde"
Sep 13 21:08:42 05[CFG]   certificate "CN=abcde" key: 1024 bit RSA
Sep 13 21:08:42 05[CFG]   using trusted ca certificate "CN=IKEv2 VPN CA, O=IKEv2 VPN"
Sep 13 21:08:42 05[CFG] checking certificate status of "CN=abcde"
Sep 13 21:08:42 05[CFG] ocsp check skipped, no ocsp found
Sep 13 21:08:42 05[CFG] certificate status is not available
Sep 13 21:08:42 05[CFG]   certificate "CN=IKEv2 VPN CA, O=IKEv2 VPN" key: 1024 bit RSA
Sep 13 21:08:42 05[CFG]   reached self-signed root ca with a path length of 0
Sep 13 21:08:42 05[IKE] authentication of '*.org' with RSA signature successful
Sep 13 21:08:42 05[ENC] added payload of type ID_RESPONDER to message
```

 **test3:** 

Python Client A ( with ID_FQDN: *.org and cert CN=abcde, SAN dns: *.org) to connect with Strongswan Server C (with config rightid=@user1.org)

**Result: AUTHENTICATION failed**

C: ipsec.conf

```
config setup

conn %default
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1

conn test
    keyexchange=ikev2
    ike=3des-sha1-modp1024!
    rekey=no
    left=192.168.0.128
    leftid=%any
    leftauth=pubkey
    leftsubnet=0.0.0.0/0
    leftcert=192.168.0.128.pem
    leftfirewall=yes
    rightsourceip=10.31.2.0/24
    right=192.168.0.1
    rightid="@user1.org"
    rightauth=pubkey
    dpdaction=clear
    fragmentation=yes
    auto=add
```

C:strongswan_user1_org.log

```
Sep 13 21:09:11 08[CFG]   using certificate "CN=abcde"
Sep 13 21:09:11 08[CFG]   certificate "CN=abcde" key: 1024 bit RSA
Sep 13 21:09:11 08[CFG]   using trusted ca certificate "CN=IKEv2 VPN CA, O=IKEv2 VPN"
Sep 13 21:09:11 08[CFG] checking certificate status of "CN=abcde"
Sep 13 21:09:11 08[CFG] ocsp check skipped, no ocsp found
Sep 13 21:09:11 08[CFG] certificate status is not available
Sep 13 21:09:11 08[CFG]   certificate "CN=IKEv2 VPN CA, O=IKEv2 VPN" key: 1024 bit RSA
Sep 13 21:09:11 08[CFG]   reached self-signed root ca with a path length of 0
Sep 13 21:09:11 08[IKE] authentication of '*.org' with RSA signature successful
Sep 13 21:09:11 08[CFG] constraint check failed: identity 'user1.org' required 
Sep 13 21:09:11 08[CFG] selected peer config 'test' unacceptable: non-matching authentication done
Sep 13 21:09:11 08[CFG] no alternative config found
Sep 13 21:09:11 08[ENC] added payload of type NOTIFY to message
```

As the test results show, the two implementations respond differently to the match of the wildcard certificate. Although IPSec supports the use of SSL / TLS certificates, I still believe that the authentication and identification policies of the two protocols should not be mixed together. 
Now, what do you think about the issue of wildcard certificates?
Thank you so much.

**Comment by colorful-sky** [comment]
Hi, paul.
I summarize the connection test of pure IPsec implementations :
Three machines participated in the test：
device A: Libreswan Client, 192.168.0.131
device B: Libreswan Server, 192.168.0.136
device C: Strongswan Server, 192.168.0.128
All test certificates are generated by Libreswan certutil. 

```
certutil -z <(head -c 1024 /dev/urandom) \
  -S -x -n "IKEv2_VPN_CA" \
  -s "O=IKEv2_VPN,CN=VPN_CA" \
  -k rsa -g 4096 -v 120 \
  -d sql:/etc/ipsec.d -t "CT,," -2

cert_name="192.168.0.131"

certutil -z <(head -c 1024 /dev/urandom) \
  -S -c "IKEv2_VPN_CA" -n "192.168.0.131" \
  -s "O=IKEv2_VPN,CN=192.168.0.131" \
  -k rsa -g 4096 -v 120 \
  -d sql:/etc/ipsec.d -t ",," \
  --keyUsage digitalSignature,keyEncipherment \
  --extKeyUsage serverAuth \
  --extSAN "dns:*.org"



cert_name="192.168.0.136"

certutil -z <(head -c 1024 /dev/urandom) \
  -S -c "IKEv2_VPN_CA" -n "192.168.0.136" \
  -s "O=IKEv2_VPN,CN=192.168.0.136" \
  -k rsa -g 4096 -v 120 \
  -d sql:/etc/ipsec.d -t ",," \
  --keyUsage digitalSignature,keyEncipherment \
  --extKeyUsage serverAuth \
  --extSAN "ip:192.168.0.136"


cert_name="192.168.0.128"

certutil -z <(head -c 1024 /dev/urandom) \
  -S -c "IKEv2_VPN_CA" -n "192.168.0.128" \
  -s "O=IKEv2_VPN,CN=192.168.0.128" \
  -k rsa -g 4096 -v 120 \
  -d sql:/etc/ipsec.d -t ",," \
  --keyUsage digitalSignature,keyEncipherment \
  --extKeyUsage serverAuth \
  --extSAN "ip:192.168.0.128"
```

Similarly, there are three tests.

#### test1: 

Libreswan Client A ( with ID_FQDN: *.org and cert O=IKEv2_VPN,CN=192.168.0.131, SAN dns: *.org) to connect with Libreswan Server B (with config rightid=@user1.org)

Result: AUTHENTICATION succeeded

A: ipsec.conf

```
version 2.0
config setup
  virtual-private=%v4:10.0.0.0/8,%v4:192.168.0.0/16,%v4:172.16.0.0/12,%v4:!192.168.42.0/24,%v4:!192.168.43.0/24
  protostack=netkey
  interfaces=%defaultroute
  uniqueids=no
  plutodebug="all crypt"
  plutostderrlog=/usr/local/etc/libreswan.log

include /etc/ipsec.d/*.conf

conn ikev2-cert_client
  left=%defaultroute
  leftcert=192.168.0.131
  leftid=@*.org
  leftsendcert=always
  
  # leftrsasigkey=%cert
  right=192.168.0.136
  rightid=%fromcert
  leftca=%same
  rightca=%same
  rightrsasigkey=%cert
  narrowing=yes
  dpddelay=30
  dpdtimeout=120
  dpdaction=clear
  auto=add
  ikev2=insist
  rekey=no
  pfs=no
  ike-frag=yes
  # ike=3des-sha1;modp1024
  phase2alg=aes_gcm-null,aes128-sha1,aes256-sha1,aes128-sha2,aes256-sha2,3des-sha1
  modecfgdns="8.8.8.8 8.8.4.4"
  encapsulation=yes
  mobike=no

  leftsubnet=192.168.0.0/24
  # rightaddresspool=0.0.0.1-255.255.255.255
  authby=rsasig
```

B: ipsec.conf

```
version 2.0
config setup
  virtual-private=%v4:10.0.0.0/8,%v4:192.168.0.0/16,%v4:172.16.0.0/12,%v4:!192.168.42.0/24,%v4:!192.168.43.0/24
  protostack=netkey
  interfaces=%defaultroute
  uniqueids=no
  plutodebug="all crypt"
  plutostderrlog=/usr/local/etc/libreswan.log

include /etc/ipsec.d/*.conf

conn ikev2-cert_server
  left=%defaultroute
  # leftcert=libswan_client_SAN_star_a_a
  # leftcert=libswan_SAN_star_a
  leftcert=192.168.0.136
  leftid=%fromcert
  leftsendcert=always
  # leftsubnet=0.0.0.0/0
  # leftrsasigkey=%cert
  # leftrsasigkey=libswan_client_SAN_star_a_a
  right=192.168.0.131
  rightid="@user1.org"
  # rightid=%fromcert
  leftca=%same
  # rightaddresspool=192.168.2.0/24
  rightca=%same
  rightrsasigkey=%cert
  narrowing=yes
  dpddelay=30
  dpdtimeout=120
  dpdaction=clear
  auto=add
  ikev2=insist
  rekey=no
  pfs=no
  ike-frag=yes
  # ike=3des-sha1;modp1024,
  phase2alg=aes_gcm-null,aes128-sha1,aes256-sha1,aes128-sha2,aes256-sha2,3des-sha1
  modecfgdns="8.8.8.8 8.8.4.4"
  encapsulation=yes
  mobike=no

  authby=rsasig
```

A.console

```
root@apple-virtual-machine:~/new_CA# ipsec restart
Redirecting to: systemctl restart ipsec.service
root@apple-virtual-machine:~/new_CA# ipsec auto --add "ikev2-cert_client"
002 added connection description "ikev2-cert_client"
root@apple-virtual-machine:~/new_CA# ipsec auto --up "ikev2-cert_client"
181 "ikev2-cert_client"[1] 192.168.0.136 #1: initiating IKEv2 IKE SA
181 "ikev2-cert_client"[1] 192.168.0.136 #1: STATE_PARENT_I1: sent v2I1, expected v2R1
182 "ikev2-cert_client"[1] 192.168.0.136 #2: STATE_PARENT_I2: sent v2I2, expected v2R2 {auth=IKEv2 cipher=AES_GCM_16_256 integ=n/a prf=HMAC_SHA2_512 group=MODP2048}
003 "ikev2-cert_client"[1] 192.168.0.136 #2: Certificate O=IKEv2_VPN,CN=192.168.0.136 failed TLS Client verification
003 "ikev2-cert_client"[1] 192.168.0.136 #2: warning: Certificate type not approved for application.
002 "ikev2-cert_client"[1] 192.168.0.136 #2: certificate verified OK: O=IKEv2_VPN,CN=192.168.0.136
002 "ikev2-cert_client"[1] 192.168.0.136 #2: IKEv2 mode peer ID is ID_DER_ASN1_DN: 'CN=192.168.0.136, O=IKEv2_VPN'
003 "ikev2-cert_client"[1] 192.168.0.136 #2: Authenticated using RSA with IKEv2_AUTH_HASH_SHA2_512
002 "ikev2-cert_client"[1] 192.168.0.136 #2: negotiated connection [192.168.0.131-192.168.0.131:0-65535 0] -> [192.168.0.136-192.168.0.136:0-65535 0]
004 "ikev2-cert_client"[1] 192.168.0.136 #2: STATE_V2_IPSEC_I: IPsec SA established tunnel mode {ESP/NAT=>0xeb90f625 <0xcf0cd4f1 xfrm=AES_GCM_16_128-NONE NATOA=none NATD=192.168.0.136:4500 DPD=active}
root@apple-virtual-machine:~/new_CA# ipsec auto --down "ikev2-cert_client"
002 "ikev2-cert_client"[1] 192.168.0.136: terminating SAs using this connection
002 "ikev2-cert_client"[1] 192.168.0.136 #2: deleting state (STATE_V2_IPSEC_I) aged 2.288s and sending notification
005 "ikev2-cert_client"[1] 192.168.0.136 #2: ESP traffic information: in=0B out=0B
002 "ikev2-cert_client"[1] 192.168.0.136 #1: deleting state (STATE_PARENT_I3) aged 2.302s and sending notification
002 "ikev2-cert_client"[1] 192.168.0.136: deleting connection "ikev2-cert_client"[1] 192.168.0.136 instance with peer 192.168.0.136 {isakmp=#0/ipsec=#0}
root@apple-virtual-machine:~/new_CA# 

```

#### test2: 

```
Libreswan Client A ( with ID_FQDN: *.org and cert O=IKEv2_VPN,CN=192.168.0.131, SAN dns: *.org) to connect with Strongswan Server C (with config rightid="@*.org")
```

Result: AUTHENTICATION succeeded

A: ipsec.conf

```
version 2.0
config setup
  virtual-private=%v4:10.0.0.0/8,%v4:192.168.0.0/16,%v4:172.16.0.0/12,%v4:!192.168.42.0/24,%v4:!192.168.43.0/24
  protostack=netkey
  interfaces=%defaultroute
  uniqueids=no
  plutodebug="all crypt"
  plutostderrlog=/usr/local/etc/libreswan.log

include /etc/ipsec.d/*.conf

conn ikev2-cert_client
  left=%defaultroute
  leftcert=192.168.0.131
  leftid=@*.org
  leftsendcert=always
  # leftrsasigkey=%cert
  right=192.168.0.128
  rightid=%fromcert
  leftca=%same
  rightca=%same
  rightrsasigkey=%cert
  narrowing=yes
  dpddelay=30
  dpdtimeout=120
  dpdaction=clear
  auto=add
  ikev2=insist
  rekey=no
  pfs=no
  ike-frag=yes
  # ike=3des-sha1;modp1024
  phase2alg=aes_gcm-null,aes128-sha1,aes256-sha1,aes128-sha2,aes256-sha2,3des-sha1
  modecfgdns="8.8.8.8 8.8.4.4"
  encapsulation=yes
  mobike=no
  leftsubnet=192.168.2.0/24
  # rightaddresspool=0.0.0.1-255.255.255.255
  authby=rsasig
```

C: ipsec.conf

```
config setup

conn %default
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1

conn test
    keyexchange=ikev2
    # ike=3des-sha1-modp1024!
    rekey=no
    left=192.168.0.128
    leftid=%any
    leftauth=pubkey
    leftcert=192.168.0.128.pem
    leftfirewall=yes
    right=192.168.0.131
    # rightid="@user1.org"
    rightid="@*.org"
    rightauth=pubkey
    dpdaction=clear
    fragmentation=yes
    auto=add
    rightsourceip=192.168.2.0/24
    # leftsubnet=192.168.43.0/8
    # leftsubnet=0.0.0.0
    # rsa_pss = yes
```

A.console

```
root@apple-virtual-machine:~/new_CA# ipsec restart
Redirecting to: systemctl restart ipsec.service
root@apple-virtual-machine:~/new_CA# ipsec auto --add "ikev2-cert_client"
002 added connection description "ikev2-cert_client"
root@apple-virtual-machine:~/new_CA# ipsec auto --up "ikev2-cert_client"
181 "ikev2-cert_client"[1] 192.168.0.128 #1: initiating IKEv2 IKE SA
181 "ikev2-cert_client"[1] 192.168.0.128 #1: STATE_PARENT_I1: sent v2I1, expected v2R1
002 "ikev2-cert_client"[1] 192.168.0.128 #1: Received unauthenticated INVALID_KE_PAYLOAD response to DH MODP2048; resending with suggested DH DH19
181 "ikev2-cert_client"[1] 192.168.0.128 #1: STATE_PARENT_I1: sent v2I1, expected v2R1
182 "ikev2-cert_client"[1] 192.168.0.128 #2: STATE_PARENT_I2: sent v2I2, expected v2R2 {auth=IKEv2 cipher=AES_CBC_256 integ=HMAC_SHA2_256_128 prf=HMAC_SHA2_256 group=DH19}
003 "ikev2-cert_client"[1] 192.168.0.128 #2: Certificate O=IKEv2_VPN,CN=192.168.0.128 failed TLS Client verification
003 "ikev2-cert_client"[1] 192.168.0.128 #2: warning: Certificate type not approved for application.
002 "ikev2-cert_client"[1] 192.168.0.128 #2: certificate verified OK: O=IKEv2_VPN,CN=192.168.0.128
002 "ikev2-cert_client"[1] 192.168.0.128 #2: IKEv2 mode peer ID is ID_DER_ASN1_DN: 'CN=192.168.0.128, O=IKEv2_VPN'
003 "ikev2-cert_client"[1] 192.168.0.128 #2: Authenticated using RSA with IKEv2_AUTH_HASH_SHA2_384
002 "ikev2-cert_client"[1] 192.168.0.128 #2: negotiated connection [192.168.2.1-192.168.2.1:0-65535 0] -> [192.168.0.128-192.168.0.128:0-65535 0]
004 "ikev2-cert_client"[1] 192.168.0.128 #2: STATE_V2_IPSEC_I: IPsec SA established tunnel mode {ESP/NAT=>0xc5759561 <0xf8ce7122 xfrm=AES_CBC_128-HMAC_SHA1_96 NATOA=none NATD=192.168.0.128:4500 DPD=active}
root@apple-virtual-machine:~/new_CA# ipsec auto --down "ikev2-cert_client"
002 "ikev2-cert_client"[1] 192.168.0.128: terminating SAs using this connection
002 "ikev2-cert_client"[1] 192.168.0.128 #2: deleting state (STATE_V2_IPSEC_I) aged 4.677s and sending notification
005 "ikev2-cert_client"[1] 192.168.0.128 #2: ESP traffic information: in=0B out=0B
002 "ikev2-cert_client"[1] 192.168.0.128 #1: deleting state (STATE_PARENT_I3) aged 4.716s and sending notification
002 "ikev2-cert_client"[1] 192.168.0.128: deleting connection "ikev2-cert_client"[1] 192.168.0.128 instance with peer 192.168.0.128 {isakmp=#0/ipsec=#0}
```

#### test3: 

Libreswan Client A ( with ID_FQDN: *.org and cert O=IKEv2_VPN,CN=192.168.0.131, SAN dns: *.org) to connect with Strongswan Server C (with config rightid="@user1.org")

Result: AUTHENTICATION failed

A: ipsec.conf  (is the same as test2 A:ipsec.conf)

C: ipsec.conf

```
config setup

conn %default
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1

conn test
    keyexchange=ikev2
    # ike=3des-sha1-modp1024!
    rekey=no
    left=192.168.0.128
    leftid=%any
    leftauth=pubkey
    leftcert=192.168.0.128.pem
    leftfirewall=yes
    right=192.168.0.131
    rightid="@user1.org"
    # rightid="@*.org"
    rightauth=pubkey
    dpdaction=clear
    fragmentation=yes
    auto=add
    rightsourceip=192.168.2.0/24
    # leftsubnet=192.168.43.0/8
    # leftsubnet=0.0.0.0
    # rsa_pss = yes
```

A.console

```
root@apple-virtual-machine:~/new_CA# ipsec restart
Redirecting to: systemctl restart ipsec.service
root@apple-virtual-machine:~/new_CA# ipsec auto --add "ikev2-cert_client"
002 added connection description "ikev2-cert_client"
root@apple-virtual-machine:~/new_CA# ipsec auto --up "ikev2-cert_client"
181 "ikev2-cert_client"[1] 192.168.0.128 #1: initiating IKEv2 IKE SA
181 "ikev2-cert_client"[1] 192.168.0.128 #1: STATE_PARENT_I1: sent v2I1, expected v2R1
002 "ikev2-cert_client"[1] 192.168.0.128 #1: Received unauthenticated INVALID_KE_PAYLOAD response to DH MODP2048; resending with suggested DH DH19
181 "ikev2-cert_client"[1] 192.168.0.128 #1: STATE_PARENT_I1: sent v2I1, expected v2R1
182 "ikev2-cert_client"[1] 192.168.0.128 #2: STATE_PARENT_I2: sent v2I2, expected v2R2 {auth=IKEv2 cipher=AES_CBC_256 integ=HMAC_SHA2_256_128 prf=HMAC_SHA2_256 group=DH19}
002 "ikev2-cert_client"[1] 192.168.0.128 #2: IKE SA authentication request rejected by peer: AUTHENTICATION_FAILED
000 "ikev2-cert_client"[1] 192.168.0.128 #2: scheduling retry attempt 1 of an unlimited number, but releasing whack
root@apple-virtual-machine:~/new_CA# 
```

C.Strongswan.log

```
Sep 15 16:51:40 02[CFG]   using certificate "CN=192.168.0.131, O=IKEv2_VPN"
Sep 15 16:51:40 02[CFG]   certificate "CN=192.168.0.131, O=IKEv2_VPN" key: 4096 bit RSA
Sep 15 16:51:40 02[CFG]   using trusted ca certificate "CN=VPN_CA, O=IKEv2_VPN"
Sep 15 16:51:40 02[CFG] checking certificate status of "CN=192.168.0.131, O=IKEv2_VPN"
Sep 15 16:51:40 02[CFG] ocsp check skipped, no ocsp found
Sep 15 16:51:40 02[CFG] certificate status is not available
Sep 15 16:51:40 02[CFG]   certificate "CN=VPN_CA, O=IKEv2_VPN" key: 4096 bit RSA
Sep 15 16:51:40 02[CFG]   reached self-signed root ca with a path length of 0
Sep 15 16:51:40 02[IKE] authentication of '*.org' with RSA_EMSA_PSS_SHA2_512_SALT_64 successful
Sep 15 16:51:40 02[CFG] constraint check failed: identity 'user1.org' required 
Sep 15 16:51:40 02[CFG] selected peer config 'test' unacceptable: non-matching authentication done
Sep 15 16:51:40 02[CFG] no alternative config found
Sep 15 16:51:40 02[IKE] processing INTERNAL_IP4_ADDRESS attribute
Sep 15 16:51:40 02[IKE] processing INTERNAL_IP4_DNS attribute
Sep 15 16:51:40 02[IKE] processing INTERNAL_IP6_ADDRESS attribute
Sep 15 16:51:40 02[IKE] processing INTERNAL_IP6_DNS attribute
Sep 15 16:51:40 02[IKE] processing INTERNAL_DNS_DOMAIN attribute
Sep 15 16:51:40 02[ENC] added payload of type NOTIFY to message
```
Have you some ideas?
Looking forward to your reply. Thanks!

**Comment by cagney** [comment]
[3.1.2](https://www.rfc-editor.org/rfc/rfc4945.html#section-3.1.2).  ID_FQDN

```
   Implementations MAY support substring, wildcard, or regular
   expression matching of the contents of ID to look up the policy in
   the SPD, and such would be a matter of local security policy
   configuration.
```

so an optional feature.  It's really up to the key issuer to decide what they are willing to sign with their root CA.
